server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSource.java

```diff
@@ -68,48 +83,30 @@ public boolean needsScores() {
     }
 
     /**
-     * Build a function prepares rounding values to be called many times.
+     * Build a function to prepare {@link Rounding}s.
      * <p>
      * This returns a {@linkplain Function} because auto date histogram will
      * need to call it many times over the course of running the aggregation.
+     * Other aggregations should feel free to call it once.
      */
     protected abstract Function<Rounding, Rounding.Prepared> roundingPreparer() throws IOException;
 
     /**
      * Check if this values source supports using global and segment ordinals.
+     * <p>
+     * If this returns {@code true} then it is safe to cast it to {@link ValuesSource.Bytes.WithOrdinals}.
      */
     public boolean hasOrdinals() {
         return false;
     }
 
-    public static class Range extends ValuesSource {
-        private final RangeType rangeType;
-        protected final IndexFieldData<?> indexFieldData;
-
-        public Range(IndexFieldData<?> indexFieldData, RangeType rangeType) {
-            this.indexFieldData = indexFieldData;
-            this.rangeType = rangeType;
-        }
-
-        @Override
-        public SortedBinaryDocValues bytesValues(LeafReaderContext context) {
-            return indexFieldData.load(context).getBytesValues();
-        }
-
-        @Override
-        public DocValueBits docsWithValue(LeafReaderContext context) throws IOException {
-            final SortedBinaryDocValues bytes = bytesValues(context);
-            return org.elasticsearch.index.fielddata.FieldData.docsWithValue(bytes);
-        }
-
-        @Override
-        public Function<Rounding, Prepared> roundingPreparer() throws IOException {
-            // TODO lookup the min and max rounding when appropriate
-            return Rounding::prepareForUnknown;
-        }
-
-        public RangeType rangeType() { return rangeType; }
-    }
+    /**
+     * {@linkplain ValuesSource} for fields who's values are best thought of
+     * as byte arrays without any other meaning. Aggregations that operate on
```

Comment:
Might be worth noting that we generally consider `keywords` to be meaningless byte arrays.  I changed the name of the `CoreValuesSourceType` from `BYTES` to `KEYWORD`, because I wanted to make it clear that it was stuff you could stick into a terms agg.  I don't think renaming here makes sense, because I expect we'll have non-keywordy bytes stuff eventually (that's probably what vectors should actually be, for example), but still, probably worth noting that this `ValuesSource` is often used for keywords.

Reply:
Done.